Add functions to properly truncate and copy UTF-8 strings

Thalley opened the pull request:
Add functions that can properly truncate and copy UTF-8 strings
without leaving unterminated started characters,
as UTF-8 characters can be 1-4 bytes long.

fixes https://github.com/zephyrproject-rtos/zephyr/issues/34159

Thalley committed:
zephyr: Add function to properly truncate UTF-8 strings

Add a function that can properly truncate UTF-8 strings
without leaving unterminated started characters,
as UTF-8 characters can be 1-4 bytes long.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
zephyr: Add UTF-8 truncating strlcpy variant

Add a function to copy a UTF-8 encoded string that
ensure correct truncation of the string if the source
is larger than the destination, as well as ensuring that
the resulting destination string is NULL-terminated.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

SebastianBoe commented on lib/os/CMakeLists.txt:
```diff
@@ -24,6 +24,7 @@ zephyr_sources(
   heap-validate.c
   bitarray.c
   multi_heap.c
+  utf8.c
```
Needs a Kconfig option to not slow down the build for all that do not use these functions.

Thalley replied:
@SebastianBoe I can absolutely add that :) 

Should have the same for some of the above as well? I figure that stuff like `printk`, `cbprintf`, `hex.c` and likely other suffer from the same issue, don't they?

SebastianBoe replied:
Yes, but it will be a backwards-incompatible change to add this after-the-fact, so it is a bit awkward. And definitely not something I expect you to do in this PR.

Also some things, that are used by core features, will be enabled so often that we might as well always enable them.

But it's difficult to identify these.

SebastianBoe reviewed (changes requested):
Changes requested

Thalley committed:
zephyr: Add UTF8 kconfig to conditionally compile utf8.c

Add the kconfig option so that the utf8.c file can be
conditionally compile, and only for the applications
that need it.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>
